pytorch_lightning/profiler/pytorch.py

```diff
@@ -521,10 +494,10 @@ def _delete_profilers(self) -> None:
             self._register = None
 
     def teardown(self, stage: Optional[str] = None) -> None:
-        self._delete_profilers()
+        for _, v in self._recording_map.items():
+            v.__exit__(None, None, None)
+        self._recording_map.clear()
 
-        for k in self._recording_map:
-            self.stop(k)
```

Comment:
Why not use `stop` anymore?

At least you can extract the `__exit__` logic inside of it to a separate protected function that `teardown` also calls.

Reply:
Because the `stop`  call the `del self._recording_map[action_name]` which can not be called when we iterate self._recording_map.
Yep, I extract the `__exit__` logit to a separate function.

Reply:
Let's rename to `_exit_and_clear_recording_map` then for readability.